runtime/runtime/src/lib.rs

```diff
@@ -1651,6 +1643,20 @@ impl Runtime {
         state_update.commit(StateChangeCause::UpdatedDelayedReceipts);
         self.apply_state_patch(&mut state_update, state_patch);
         let (trie, trie_changes, state_changes) = state_update.finalize()?;
+        if let Some(prefetcher) = &prefetcher {
+            // Only clear the prefetcher queue after finalize is done because as part of receipt
+            // processing we also prefetch account data and access keys that are accessed in
+            // finalize. This data can take a very long time otherwise if not prefetched.
```

Comment:
Actually, if we follow the initial design of the prefetcher, this means we are just sending those prefetch requests too early. The prefetches are made before processing but we need the values after processing.

So, what I would have tried first is this: Instead of prefetching the access keys and account data in the `prefetch_receipt_data()` call, I would have added a new method to the prefetcher. Maybe something like `prefetch_finalize` where we specifically read data that was written but never read. Sure, we start it a bit later, but it's still with 8 background threads + 1 main thread instead of just the main thread.
Then I could still do the clearing in all places but guarantee parallel fetches on the data for finalize.

But what you have seems to work good enough, and I'm not sure my approach would actually give equal or better results.

Reply:
I'm happy to explore this idea… as a follow up. Thank you for writing it down :)